File: living_world/queries.py
"""Read-only world queries — used by dashboard, CLI, future API.

Foundation step of the dashboard 3-layer split (KNOWN_ISSUES #3).

Today the dashboard inlines a lot of "filter events / count things /
group by pack" code. By moving those helpers here, ANY consumer (the
current Streamlit dashboard, a future Svelte one, a Jupyter notebook,
a CLI report) reuses the same logic with the same field names.

Discipline: this module never mutates state. Pure functions over
World / TickEngine snapshots. Anything that *changes* the world
belongs in `living_world/phases.py` or `rules/*.py` instead.
"""

from __future__ import annotations

from collections import Counter
from dataclasses import dataclass

from living_world.core.event import LegendEvent
from living_world.core.world import World


# ── Event queries ──────────────────────────────────────────────────────────

def recent_events(world: World, n: int = 50) -> list[LegendEvent]:
    """Last N events across all packs, oldest-first."""
    all_evts = world.events_since(1)
    return all_evts[-n:]


def events_by_pack(world: World, since_tick: int = 1) -> dict[str, list[LegendEvent]]:
    """Group events by pack_id — useful for per-pack chronicle panels."""
    out: dict[str, list[LegendEvent]] = {}
    for e in world.events_since(since_tick):
        out.setdefault(e.pack_id, []).append(e)
    return out


def events_by_day(world: World, since_tick: int = 1) -> dict[int, list[LegendEvent]]:
    """Group events by tick — used by the digest CLI."""
    out: dict[int, list[LegendEvent]] = {}
    for e in world.events_since(since_tick):
        out.setdefault(e.tick, []).append(e)
    return out


def event_kind_distribution(
    world: World, since_tick: int = 1, *, top_k: int = 10,
) -> list[tuple[str, int]]:
    """Top-K most frequent event kinds since `since_tick`."""
    c: Counter = Counter(e.event_kind for e in world.events_since(since_tick))
    return c.most_common(top_k)


def diversity_summary(world: World, since_tick: int = 1) -> dict:
    """Headline diversity stats: total / unique / top-kind percent."""
    evts = world.events_since(since_tick)
    if not evts:
        return {"total": 0, "unique": 0, "top_kind": None, "top_pct": 0.0}
    kinds = Counter(e.event_kind for e in evts)
    top = kinds.most_common(1)[0]
    return {
        "total": len(evts),
        "unique": len(kinds),
        "top_kind": top[0],
        "top_pct": 100 * top[1] / len(evts),
    }


# ── Agent queries ──────────────────────────────────────────────────────────

def living_agents(world: World, pack_id: str | None = None) -> list:
    return [
        a for a in world.living_agents()
        if pack_id is None or a.pack_id == pack_id
    ]


def death_count(world: World, pack_id: str | None = None) -> int:
    return sum(
        1 for a in world.all_agents()
        if not a.is_alive() and (pack_id is None or a.pack_id == pack_id)
    )


# ── Engine status (for dashboard "what's wired" panel) ─────────────────────

@dataclass
class FeatureStatus:
    name: str
    on: bool
    detail: str = ""


def feature_status(engine) -> list[FeatureStatus]:
    """Compact list of LLM features that are wired and their settings.

    Powers the dashboard's "what's actually live this run" sidebar.
    """
    out = [
        FeatureStatus("memory",
                       engine.memory is not None,
                       f"reflect_every={engine.reflect_every_ticks}"),
        FeatureStatus("planner",
                       engine.agent_planner is not None,
                       f"hf_only={engine.plan_hf_only}"),
        FeatureStatus("chronicler",
                       engine.chronicler is not None,
                       f"every={engine.chronicle_every_ticks}t"),
        FeatureStatus("emergent",
                       engine.emergent_proposer is not None,
                       f"max_per_tick={engine.emergent_max_per_tick}"),
        FeatureStatus("conscience", engine.consciousness is not None),
        FeatureStatus("dialogue_loop", engine.conversation_loop_enabled),
        FeatureStatus("perception",
                       engine.perception is not None,
                       f"≥{engine.perception_threshold:.2f}"
                       if engine.perception is not None else ""),
        FeatureStatus("self_update",
                       engine.self_update is not None,
                       f"≥{engine.self_update_threshold:.2f}"
                       if engine.self_update is not None else ""),
        FeatureStatus("llm_movement",
                       engine.movement.llm_advisor is not None,
                       f"chance={engine.movement.llm_chance:.2f}"
                       if engine.movement.llm_advisor is not None else ""),
        FeatureStatus("memory_reflector",
                       engine.memory is not None
                       and engine.memory.reflector is not None),
    ]
    return out


def narrator_stats(engine) -> dict:
    s = engine.narrator.stats
    return {"tier1": s.tier1, "tier3": s.tier3, "total": s.tier1 + s.tier3}


# ── Chronicle export ──────────────────────────────────────────────────────

def export_chronicle_markdown(world: World) -> str:
    """Render world.chapters as a Markdown chronicle.

    Used by `lw export-chronicle` and any future "share my run" feature.
    Pure function — no I/O, no side effects.
    """
    if not world.chapters:
        return "# (no chapters yet)\n\nThe chronicler hasn't fired in this run.\n"

    by_pack: dict[str, list[dict]] = {}
    for ch in world.chapters:
        by_pack.setdefault(ch.get("pack_id", "?"), []).append(ch)

    lines: list[str] = ["# Chronicle", ""]
    lines.append(f"_World ran {world.current_tick} tick(s). "
                 f"{len(world.chapters)} chapters across "
                 f"{len(by_pack)} pack(s)._\n")

    for pack_id in sorted(by_pack.keys()):
        lines.append(f"## {pack_id}\n")
        for ch in sorted(by_pack[pack_id], key=lambda c: c.get("tick", 0)):
            title = ch.get("title", "(untitled)")
            tick = ch.get("tick", "?")
            body = ch.get("body", "").strip()
            lines.append(f"### Day {tick} — {title}\n")
            lines.append(body)
            lines.append("")  # blank between chapters
    return "\n".join(lines)

File: packages/sim-core/src/queries.ts
/**
 * Read-only world queries — TypeScript port of `living_world.queries`.
 *
 * Pure data-in / data-out: caller supplies the event list (or chapter
 * list); we never reach into a `World` object. This keeps the module
 * portable to the browser, where `World` doesn't exist.
 *
 * Phase 3 motivation:
 *   These helpers were called from the FastAPI server PER REQUEST.
 *   The dashboard now runs them client-side from a single `/api/events`
 *   fetch, eliminating round-trips for every filter / aggregation toggle.
 *
 * Parity discipline (AI-Native criterion A):
 *   `scripts/dump_fixtures.py` writes JSON fixtures from Python; this
 *   module's tests assert byte-identical results. Any algorithmic drift
 *   is caught at `make check` time.
 */

import type { Outcome } from "./dice.js";

// ── DTOs (mirror the Python LegendEvent / Chapter shape) ────────────────

/**
 * Event projection used by the query helpers. Wider than `EventLite` in
 * dice.ts because queries need pack/tier metadata that dice doesn't.
 * Use this for inputs to query functions; dashboard receives matching
 * `WorldEvent` from `/api/events`.
 */
export interface EventForQueries {
  eventId: string;
  tick: number;
  packId: string;
  tileId: string;
  eventKind: string;
  outcome: Outcome;
  importance: number;
  tierUsed: number;
  isEmergent: boolean;
  participants: ReadonlyArray<string>;
  // Optional rendering fields the chronicle export reads.
  templateRendering?: string;
  spotlightRendering?: string;
}

export interface ChapterForExport {
  tick: number;
  packId: string;
  title: string;
  body: string;
  eventIds: ReadonlyArray<string>;
}

// ── Event list slicers ──────────────────────────────────────────────────

/**
 * Last N events (oldest first) — caller passes the full ordered list.
 * Returns the same array slice Python's `recent_events` would.
 */
export function recentEvents(
  events: ReadonlyArray<EventForQueries>,
  n: number = 50,
): EventForQueries[] {
  if (n >= events.length) return events.slice();
  return events.slice(events.length - n);
}

/** Events grouped by `packId`. Insertion order preserved per pack. */
export function eventsByPack(
  events: ReadonlyArray<EventForQueries>,
): Map<string, EventForQueries[]> {
  const out = new Map<string, EventForQueries[]>();
  for (const e of events) {
    const arr = out.get(e.packId);
    if (arr) arr.push(e);
    else out.set(e.packId, [e]);
  }
  return out;
}

/** Events grouped by tick. Tick keys come back sorted ascending. */
export function eventsByDay(
  events: ReadonlyArray<EventForQueries>,
): Map<number, EventForQueries[]> {
  const out = new Map<number, EventForQueries[]>();
  for (const e of events) {
    const arr = out.get(e.tick);
    if (arr) arr.push(e);
    else out.set(e.tick, [e]);
  }
  // Re-sort by key for stable consumer iteration; Python's defaultdict
  // doesn't guarantee key order across insertion patterns.
  return new Map([...out.entries()].sort((a, b) => a[0] - b[0]));
}

// ── Aggregations ────────────────────────────────────────────────────────

/**
 * Top-K most frequent event kinds, descending by count, ties broken by
 * insertion order (matching Python `collections.Counter.most_common`).
 */
export function eventKindDistribution(
  events: ReadonlyArray<EventForQueries>,
  topK: number = 10,
): Array<[string, number]> {
  const counts = new Map<string, number>();
  for (const e of events) {
    counts.set(e.eventKind, (counts.get(e.eventKind) ?? 0) + 1);
  }
  // Stable sort: Map iteration is insertion order, sort is stable in V8.
  return [...counts.entries()]
    .sort((a, b) => b[1] - a[1])
    .slice(0, topK);
}

export interface DiversitySummary {
  total: number;
  unique: number;
  top_kind: string | null;
  top_pct: number;
}

/**
 * Headline diversity stats. Field names use snake_case for the two
 * legacy keys (top_kind / top_pct) so this stays compatible with the
 * Python output that ships through `/api/world` today.
 */
export function diversitySummary(
  events: ReadonlyArray<EventForQueries>,
): DiversitySummary {
  if (events.length === 0) {
    return { total: 0, unique: 0, top_kind: null, top_pct: 0.0 };
  }
  const counts = new Map<string, number>();
  for (const e of events) counts.set(e.eventKind, (counts.get(e.eventKind) ?? 0) + 1);
  // most_common(1)
  let topKind = "";
  let topN = 0;
  for (const [k, v] of counts) {
    if (v > topN) {
      topKind = k;
      topN = v;
    }
  }
  return {
    total: events.length,
    unique: counts.size,
    top_kind: topKind,
    top_pct: (100 * topN) / events.length,
  };
}

// ── Chronicle markdown export ───────────────────────────────────────────

/**
 * Render the chapter list as a Markdown chronicle.
 *
 * Mirrors `living_world.queries.export_chronicle_markdown` exactly so
 * "Download chronicle.md" can run in the browser without a round-trip.
 */
export function exportChronicleMarkdown(
  chapters: ReadonlyArray<ChapterForExport>,
  currentTick: number,
): string {
  if (chapters.length === 0) {
    return "# (no chapters yet)\n\nThe chronicler hasn't fired in this run.\n";
  }
  const byPack = new Map<string, ChapterForExport[]>();
  for (const ch of chapters) {
    const arr = byPack.get(ch.packId);
    if (arr) arr.push(ch);
    else byPack.set(ch.packId, [ch]);
  }
  const lines: string[] = ["# Chronicle", ""];
  lines.push(
    `_World ran ${currentTick} tick(s). ${chapters.length} chapters across ` +
      `${byPack.size} pack(s)._\n`,
  );
  // Sort packs alphabetically; chapters within each pack by tick asc.
  for (const packId of [...byPack.keys()].sort()) {
    lines.push(`## ${packId}\n`);
    const pack = byPack.get(packId)!;
    const sorted = [...pack].sort((a, b) => (a.tick ?? 0) - (b.tick ?? 0));
    for (const ch of sorted) {
      const title = ch.title || "(untitled)";
      const tick = ch.tick;
      const body = (ch.body ?? "").trim();
      lines.push(`### Day ${tick} — ${title}\n`);
      lines.push(body);
      lines.push(""); // blank between chapters
    }
  }
  return lines.join("\n");
}
